bee-tangle/src/metadata.rs

```diff
@@ -0,0 +1,304 @@
+        self.flags.set_referenced(true);
+        self.reference_timestamp = timestamp;
+    }
+
+    /// Get the conflict state of this message.
+    pub fn conflict(&self) -> ConflictReason {
+        self.conflict
+    }
+
+    /// Set the conflict state of this message.
+    pub fn set_conflict(&mut self, conflict: ConflictReason) {
+        self.conflict = conflict;
+    }
+}
+
+/// An error that may occur when manipulating message metadata.
+#[derive(Debug)]
+pub enum MessageMetadataError {
+    /// An IO error occurred.
+    Io(std::io::Error),
+    /// A packing error occurred.
+    OptionIndex(<Option<MilestoneIndex> as Packable>::Error),
+    /// A packing error occurred.
+    OptionIndexId(<Option<IndexId> as Packable>::Error),
+}
+
+impl From<std::io::Error> for MessageMetadataError {
+    fn from(error: std::io::Error) -> Self {
+        MessageMetadataError::Io(error)
+    }
+}
+
+impl From<OptionError<std::io::Error>> for MessageMetadataError {
+    fn from(error: OptionError<std::io::Error>) -> Self {
+        MessageMetadataError::OptionIndex(error)
+    }
+}
+
+impl From<OptionError<IndexIdError>> for MessageMetadataError {
+    fn from(error: OptionError<IndexIdError>) -> Self {
+        MessageMetadataError::OptionIndexId(error)
+    }
+}
+
+impl Packable for MessageMetadata {
+    type Error = MessageMetadataError;
+
+    fn packed_len(&self) -> usize {
+        self.flags.packed_len()
+            + self.milestone_index.packed_len()
+            + self.arrival_timestamp.packed_len()
+            + self.solidification_timestamp.packed_len()
+            + self.reference_timestamp.packed_len()
+            + self.omrsi.packed_len()
+            + self.ymrsi.packed_len()
+            + 0u8.packed_len()
+    }
+
+    fn pack<W: Write>(&self, writer: &mut W) -> Result<(), Self::Error> {
+        self.flags.pack(writer)?;
+        self.milestone_index.pack(writer)?;
+        self.arrival_timestamp.pack(writer)?;
+        self.solidification_timestamp.pack(writer)?;
+        self.reference_timestamp.pack(writer)?;
+        self.omrsi.pack(writer)?;
+        self.ymrsi.pack(writer)?;
+        (self.conflict as u8).pack(writer)?;
+
+        Ok(())
+    }
+
+    fn unpack_inner<R: Read + ?Sized, const CHECK: bool>(reader: &mut R) -> Result<Self, Self::Error> {
+        Ok(Self {
+            flags: Flags::unpack_inner::<R, CHECK>(reader)?,
+            milestone_index: Option::<MilestoneIndex>::unpack_inner::<R, CHECK>(reader)?,
+            arrival_timestamp: u64::unpack_inner::<R, CHECK>(reader)?,
+            solidification_timestamp: u64::unpack_inner::<R, CHECK>(reader)?,
+            reference_timestamp: u64::unpack_inner::<R, CHECK>(reader)?,
+            omrsi: Option::<IndexId>::unpack_inner::<R, CHECK>(reader)?,
+            ymrsi: Option::<IndexId>::unpack_inner::<R, CHECK>(reader)?,
+            conflict: ConflictReason::unpack_inner::<R, CHECK>(reader)?,
+        })
+    }
+}
+
+/// An identifier for message indices.
```

Comment:
Since I added this type a long time ago: A more accurate description would probably be:
"A type that is used to associate two particular interesting Cone Root Indexes with a message in the Tangle, i.e. the Oldest Cone Root Index (OCRI), and the Youngest Cone Root Index (YCRI)"
One merged I probably open a PR to suggest the renaming of this type to "ConeRootIndex".

Reply:
It's not clear to me: what does 'interesting' mean in this context?

Reply:
"interesting" here means, that only two particular of the many "CRI"s are worth considering when it comes to tip selection. If you imagine a Cone with a particular milestone index at its root, then the only two interesting messages when you want to attach this cone to a previous one (in order to be accepted by the network) are the ones consisting of:
(a) the one message that is attached to the closest previous milestone (in terms of its index)
(b) the one message that is attached to the furthest previous milestone (in terms of its index)
Hence, only those two messages are interesting.

Reply:
Okay, but what do I put in the doc comment? I'm not an expert on this area, so you're going to be the better one to come up with a comment, unless you think the first one is sufficient?

Reply:
"A type that is used to associate two particular Cone Root Indexes with a message in the Tangle, i.e. the Oldest Cone Root Index (OCRI), and the Youngest Cone Root Index (YCRI) to decide on the eligibility of a message as a tip."

Reply:
But since we are still using the old terms atm throughout the codebase, pls replace OCRI->OMRSI and YCRI->YMRSI for now. We'll change that later! :smile: